### PR TITLE
Follow-up on try/catch/finally implementation

### PR DESCRIPTION
**Pull request by AlekseyTs**
- Split RegionKind.Handler into RegionKind.Catch and RegionKind.Finally
- Improve control flow graph packing - eliminate empty finally.

### CODE REVIEW

**Review by gafter — APPROVED** [review]
:shipit:

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.RegionBuilder.cs [hunk 1]
```diff
@@ -77,19 +77,82 @@ public void Add(RegionBuilder region)
...
                         break;
                 }
 #endif
             }
 
+            public void Remove(RegionBuilder region)
+            {
+                Debug.Assert(region.Enclosing == this);
+
+                if (Regions.Count == 1)
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.RegionBuilder.cs]: Consider removing the special cases for `Regions.Count == 1` (here and in `ReplaceRegion`) unless the behavior is different from `> 1` case.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.RegionBuilder.cs]: > Consider removing the special cases for Regions.Count == 1 (here and in ReplaceRegion) unless the behavior is different from > 1 case. 

I think this optimization is very simple and worth keeping. The code path is covered by unit-tests.

---
In reply to: [171017423](https://github.com/dotnet/roslyn/pull/25079#discussion_r171017423) [](ancestors = 171017423)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -303,6 +314,49 @@ private static bool PackBlocks(ArrayBuilder<BasicBlock> blocks, PooledDictionary
...
                         Debug.Assert(currentRegion.FirstBlock == block);
+                        Debug.Assert(!currentRegion.HasRegions);
+
+                        // Remove Try/Finally if Finally is empty
+                        if (currentRegion.Kind == ControlFlowGraph.RegionKind.Finally &&
+                            block.Statements.IsEmpty && block.InternalConditional.Condition == null &&
+                            block.InternalNext.Destination == null && block.InternalNext.Flags == BasicBlock.BranchFlags.StructuredExceptionHandling &&
+                            block.Predecessors.IsEmpty)
+                        {
+                            // Nothing usefull is happening in this finally, let's remove it
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Typo: `useful`

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -249,6 +221,45 @@ bool PackRegion(RegionBuilder region)
             }
         }
 
+        private static void MergeSubRegionAndFree(RegionBuilder enclosing, RegionBuilder subRegion, ArrayBuilder<BasicBlock> blocks, PooledDictionary<BasicBlock, RegionBuilder> regionMap)
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider removing `enclosing` if it's redundant.

**Review by cston — APPROVED** [review]

**Review by mavasani — APPROVED** [review]

**Review by jinujoseph — APPROVED** [review]

**Comment by AlekseyTs** [comment]
Retest this please

**Comment by AlekseyTs** [comment]
Test this please

**Comment by 333fred** [comment]
@dotnet-bot retest this please

**Pull request merged**